pkg/validations/cluster.go

```diff
@@ -114,3 +118,132 @@ func ValidateManagementClusterBundlesVersion(ctx context.Context, k KubectlClien
+	}
+
+	parsedClusterVersion, err := semver.New(string(*v))
+	if err != nil {
+		return fmt.Errorf("parsing cluster eksa version: %v", err)
+	}
+
+	uv := workload.Cluster.Spec.EksaVersion
+	if uv == nil {
+		return fmt.Errorf("upgrade cluster version cannot be nil")
+	}
+
+	parsedUpgradeVersion, err := semver.New(string(*uv))
+	if err != nil {
+		return fmt.Errorf("parsing upgrade cli version: %v", err)
+	}
+
+	majorVersionDifference := math.Abs(float64(parsedUpgradeVersion.Major) - float64(parsedClusterVersion.Major))
+	minorVersionDifference := float64(parsedUpgradeVersion.Minor) - float64(parsedClusterVersion.Minor)
+	supportedMinorVersionIncrement := float64(1)
+
+	if majorVersionDifference > 0 || !(minorVersionDifference <= supportedMinorVersionIncrement && minorVersionDifference >= 0) {
+		msg := fmt.Sprintf("WARNING: version difference between upgrade version (%d.%d) and cluster version (%d.%d) do not meet the supported version increment of +%f",
+			parsedUpgradeVersion.Major, parsedUpgradeVersion.Minor, parsedClusterVersion.Major, parsedClusterVersion.Minor, supportedMinorVersionIncrement)
+		return fmt.Errorf(msg)
+	}
+
+	return nil
+}
+
+func getVersionFromBundle(ctx context.Context, k KubectlClient, br *v1alpha1.BundlesRef, cluster *types.Cluster) (*v1alpha1.EksaVersion, error) {
+	if br == nil {
+		return nil, fmt.Errorf("bundlesRef and eksa version cannot be nil")
+	}
+
+	name := br.Name
+	ns := br.Namespace
+	bundle, err := k.GetBundles(ctx, cluster.KubeconfigFile, name, ns)
+	if err != nil {
+		return nil, err
+	}
+
+	uri := bundle.Spec.VersionsBundles[0].Eksa.ClusterController.URI
+	if !strings.Contains(uri, ":") {
+		return nil, fmt.Errorf("could not find tag in Eksa Cluster Controller Image")
+	}
+
+	tag := strings.Split(uri, ":")[1]
+	v := strings.Split(tag, "-")[0]
+	version := v1alpha1.EksaVersion(v)
+	return &version, nil
+}
+
+// ValidateManagementClusterEksaVersion ensures workload cluster isn't created by a newer version than management cluster.
+func ValidateManagementClusterEksaVersion(ctx context.Context, k KubectlClient, mgmtCluster *types.Cluster, workload *cluster.Spec) error {
+	cluster, err := k.GetEksaCluster(ctx, mgmtCluster, mgmtCluster.Name)
+	if err != nil {
+		return err
+	}
+
+	v := cluster.Spec.EksaVersion
+	if v == nil {
+		return fmt.Errorf("mgmt cluster eksaVersion cannot be nil")
+	}
+
+	mVersion, err := semver.New(string(*v))
+	if err != nil {
+		return fmt.Errorf("management cluster eksaVersion is invalid: %w", err)
+	}
+
+	wv := workload.Cluster.Spec.EksaVersion
+	if wv == nil {
+		return fmt.Errorf("workload cluster eksaVersion cannot be nil")
+	}
+
+	wVersion, err := semver.New(string(*workload.Cluster.Spec.EksaVersion))
+	if err != nil {
+		return fmt.Errorf("workload cluster eksaVersion is invalid: %w", err)
+	}
+
+	if wVersion.GreaterThan(mVersion) {
+		return fmt.Errorf("cannot upgrade workload cluster with version %v while management cluster is an older version %v", wVersion, mVersion)
+	}
```

Comment:
it also seems like this logic is duplicated in the controller
any chance we can reuse the code?